activerecord/lib/active_record/connection_adapters/abstract/schema_statements.rb

```diff
@@ -1558,8 +1559,44 @@ def create_alter_table(name)
           AlterTable.new create_table_definition(name)
         end
 
+        VALID_TABLE_OPTIONS = [:temporary, :if_not_exists, :options, :as, :comment, :charset, :collation]
+
+        def valid_table_definition_option_keys
+          VALID_TABLE_OPTIONS
+        end
+
+        VALID_PRIMARY_KEY_OPTIONS = [:limit, :default, :precision]
+
+        def valid_primary_key_option_keys
+          VALID_PRIMARY_KEY_OPTIONS
+        end
+
+        VALID_INTERNAL_OPTIONS = [:_uses_legacy_table_name, :_skip_validate_options_keys]
+
+        def validate_create_table_options!(options)
+          unless options[:_skip_validate_options_keys]
+            options.except(*VALID_INTERNAL_OPTIONS).assert_valid_keys(*(valid_table_definition_option_keys + valid_primary_key_option_keys))
+          end
+        end
+
+        def extract_primary_key_options!(options)
+          primary_key_options = options.extract!(*valid_primary_key_option_keys)
+
+          if options[:_skip_validate_options_keys]
+            primary_key_options[:_skip_validate_options_keys] = options[:_skip_validate_options_keys]
+          end
+
+          primary_key_options
+        end
```

Comment:
Wouldn't all this be much simpler if `:_skip_validate_options_keys` was part of `valid_primary_key_option_keys` etc? I don't really see a downside in doing this.

Reply:
My thinking when doing so was that I wanted to avoid exposing an internal implementation concern in the error message when an invalid option is used. Therefore the trade off I made here was to have a cleaner error message at the expense of a more verbose private method.

Reply:
I see, but that's only for `assert_valid_keys`., you can still simplify the above in a one-liner:

```ruby
options.extract!(*valid_primary_key_option_keys, :_skip_validate_options_keys)
```

So a method is likely not needed.

Reply:
Just to clarify, do we agree that it'll be nicer to avoid exposing the internal options that are being used here? My question here is kind of related to the code suggestion in https://github.com/rails/rails/pull/46178#discussion_r989139368.

Reply:
Yes we agree.